src/ReverseProxy/Service/SessionAffinity/BaseSessionAffinityProvider.cs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using Microsoft.ReverseProxy.RuntimeModel;
+
+namespace Microsoft.ReverseProxy.Service.SessionAffinity
+{
+    internal abstract class BaseSessionAffinityProvider<T> : ISessionAffinityProvider
+    {
+        protected static readonly object AffinityKeyId = new object();
+
+        public abstract string Mode { get; }
+
+        public virtual void AffinitizeRequest(HttpContext context, BackendConfig.BackendSessionAffinityOptions options, DestinationInfo destination)
+        {
+            if (!options.Enabled)
+            {
+                return;
+            }
+
+            if (!context.Items.TryGetValue(AffinityKeyId, out var affinityKey)) // If affinity key is already set on request, we assume that passed destination always matches to that key
+            {
+                affinityKey = GetDestinationAffinityKey(destination);
+            }
+
+            var encryptedKey = (string)affinityKey; // TBD. The affinity key must be encrypted.
+            SetEncryptedAffinityKey(context, options, encryptedKey);
```

Comment:
Encryption is probably an implementation detail of the provider. E.g. Cookies allow a more restricted set of characters than a general header.
```suggestion
            SetAffinityKey(context, options, affinityKey);
```